src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs

```diff
@@ -0,0 +1,790 @@
+                    writer.WriteString(value.Key, value.Value);
+                }
+                writer.WriteEndObject();
+            }
+        }
+
+        private void WriteCompletionMessage(CompletionMessage message, ref Utf8JsonWriter writer)
+        {
+            WriteInvocationId(message, ref writer);
+            if (!string.IsNullOrEmpty(message.Error))
+            {
+                writer.WriteString(ErrorPropertyNameBytes, message.Error);
+            }
+            else if (message.HasResult)
+            {
+                var bytes = JsonSerializer.ToBytes(message.Result, message.Result?.GetType());
+                using var token = JsonDocument.Parse(bytes);
+                token.RootElement.WriteAsProperty(ResultPropertyNameBytes, ref writer);
+            }
+        }
+
+        private void WriteCancelInvocationMessage(CancelInvocationMessage message, ref Utf8JsonWriter writer)
+        {
+            WriteInvocationId(message, ref writer);
+        }
+
+        private void WriteStreamItemMessage(StreamItemMessage message, ref Utf8JsonWriter writer)
+        {
+            WriteInvocationId(message, ref writer);
+
+            var bytes = JsonSerializer.ToBytes(message.Item, message.Item?.GetType());
+            using var token = JsonDocument.Parse(bytes);
+            token.RootElement.WriteAsProperty(ItemPropertyNameBytes, ref writer);
+        }
+
+        private void WriteInvocationMessage(InvocationMessage message, ref Utf8JsonWriter writer)
+        {
+            WriteInvocationId(message, ref writer);
+            writer.WriteString(TargetPropertyNameBytes, message.Target);
+
+            WriteArguments(message.Arguments, ref writer);
+
+            WriteStreamIds(message.StreamIds, ref writer);
+        }
+
+        private void WriteStreamInvocationMessage(StreamInvocationMessage message, ref Utf8JsonWriter writer)
+        {
+            WriteInvocationId(message, ref writer);
+            writer.WriteString(TargetPropertyNameBytes, message.Target);
+
+            WriteArguments(message.Arguments, ref writer);
+
+            WriteStreamIds(message.StreamIds, ref writer);
+        }
+
+        private void WriteCloseMessage(CloseMessage message, ref Utf8JsonWriter writer)
+        {
+            if (message.Error != null)
+            {
+                writer.WriteString(ErrorPropertyNameBytes, message.Error);
+            }
+        }
+
+        private void WriteArguments(object[] arguments, ref Utf8JsonWriter writer)
+        {
+            writer.WriteStartArray(ArgumentsPropertyNameBytes);
+            foreach (var argument in arguments)
+            {
+                var type = argument?.GetType();
+                if (type == typeof(DateTime))
+                {
+                    writer.WriteStringValue((DateTime)argument);
+                }
+                else if (type == typeof(DateTimeOffset))
+                {
+                    writer.WriteStringValue((DateTimeOffset)argument);
```

Comment:
Was this needed only because `+` is being escaped in the writer by default if written out as string (which JsonElement ends up doing)?

If there was a way to customize the JsonWriter to not escape `+`, for example, would you apply that setting to the writer instance you are using here for all the input data you want to write? I was considering providing a writer option to provide custom `TextEncoder`.

cc @bartonjs, @GrabYourPitchforks

Reply:
>Was this needed only because + is being escaped in the writer by default if written out as string (which JsonElement ends up doing)?

Correct. Also it's probably more performant to write it directly with the writer.